arxiv-opensearch-cohere/README.md

```diff
@@ -0,0 +1,256 @@
+* `method.engine` = the library to use for indexing/search. When using a CPU, `nmslib` is the recommended engine option
+
+When selecting `hnsw` as the `method.name`, we have additional parameters for the `hnsw` algorithm such as `ef_construction` and `m`. See docs [here](https://opensearch.org/docs/2.5/search-plugins/knn/index/) for guidance on setting the right parameters.
+
+
+
+We are using the Python [opensearch-py](https://pypi.org/project/opensearch-py/) package to communicate with the OpenSearch cluster in the backend. For demo purposes, we are turning off SSL verification since it is a simple, local cluster. 
+Create a client like so: 
+```python
+from opensearchpy import OpenSearch
+
+def get_opensearch_client(host="localhost", port=9200) -> OpenSearch:
+    # Create the client with SSL/TLS and hostname verification disabled.
+    client = OpenSearch(
+        hosts=[{"host": host, "port": port}],
+        http_compress=True,  # enables gzip compression for request bodies
+        use_ssl=False,
+        verify_certs=False,
+        ssl_assert_hostname=False,
+        ssl_show_warn=False,
+    )
+    return client
+
+client = get_opensearch_client()
+```
+
+Create an index by first creating the `body` payload and then submitting that to the index endpoint. The `body` payload specifies the name of your vectors, the size and various Approximate k-NN parameters discussed above. In this example, our document index will be called `arxiv-l2`. 
+
+```python
+INDEX_NAME = "arxiv-l2"
+
+body = {
+    "settings": {"index": {"knn": "true", "knn.algo_param.ef_search": 100}},
+    "mappings": {
+        "properties": {
+            VECTOR_NAME: {
+                "type": "knn_vector",
+                "dimension": VECTOR_SIZE,
+                "method": {
+                    "name": "hnsw",
+                    "space_type": "l2",
+                    "engine": "nmslib",
+                    "parameters": {"ef_construction": 128, "m": 24},
+                },
+            },
+        }
+    },
+}
+response = client.indices.create(INDEX_NAME, body=body)
+```
+You can sanity check your index has been created by running the following line to get all indices currently populated in your OpenSearch backend. 
+```python
+print(client.indices.get_alias("*").keys())
+```
+
+Now that the index is created, we need to populate it with data. We are going to use the `cache.jsonl` we created in the previous step and the dataset file to populate the index with documents and their corresponding embedding vectors: 
+```python
+with open("cache.jsonl", "r") as fp:
+    cache = json.load(fp)
+
+# insert each row one-at-a-time to the document index
+for i, row in tqdm(df.iterrows()):
```

Comment:
some of the code has imports, some does not.  not a big deal buy maybe remve tqdm here